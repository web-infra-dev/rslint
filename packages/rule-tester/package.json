{
  "name": "@typescript-eslint/rule-tester",
  "version": "0.1.4",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "private": true,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "default": "./dist/"
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "build": "tsc -b tsconfig.build.json --force",
    "dev": "tsc -b tsconfig.build.json --force --watch",
    "test": "rstest run --testTimeout=0",
    "test:update": "rstest run -u --testTimeout=0"
  },
  "files": [
    "dist"
  ],
  "keywords": [
    "rslint",
    "rule-tester",
    "testing",
    "linter"
  ],
  "author": "",
  "license": "MIT",
  "description": "Rule testing utilities for rslint",
  "devDependencies": {
    "@types/node": "24.0.14",
    "typescript": "5.8.3"
  },
  "dependencies": {
    "@rslint/core": "workspace:*"
  },
  "publishConfig": {
    "access": "public"
  }
}
