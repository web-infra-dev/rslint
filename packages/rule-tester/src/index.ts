// Forked and modified from https://github.com/typescript-eslint/typescript-eslint/blob/16c344ec7d274ea542157e0f19682dd1930ab838/packages/rule-tester/src/RuleTester.ts#L4

import path from 'node:path';
import { test, describe, expect } from '@rstest/core';
import { lint, type Diagnostic } from '@rslint/core';
import assert from 'node:assert';

interface TsDiagnostic {
  line: number;
  column: number;
  endLine: number;
  endColumn: number;
  messageId: string;
  suggestions: any[];
}
function toCamelCase(name: string): string {
  return name.replace(/-([a-z])/g, g => g[1].toUpperCase());
}
// check whether rslint diagnostics and typescript-eslint diagnostics are semantic equal
function checkDiagnosticEqual(
  rslintDiagnostic: Diagnostic[],
  tsDiagnostic: TsDiagnostic[],
) {
  assert(
    rslintDiagnostic.length === tsDiagnostic.length,
    `Length mismatch: ${rslintDiagnostic.length} !== ${tsDiagnostic.length}`,
  );
  for (let i = 0; i < rslintDiagnostic.length; i++) {
    const rslintDiag = rslintDiagnostic[i];
    const tsDiag = tsDiagnostic[i];
    // check rule match
    assert(
      toCamelCase(rslintDiag.messageId) === tsDiag.messageId,
      `Message mismatch: ${rslintDiag.messageId} !== ${tsDiag.messageId}`,
    );

    // check range match
    // tsDiag sometimes doesn't have line and column, so we need to check that
    if (tsDiag.line) {
      assert(
        rslintDiag.range.start.line === tsDiag.line,
        `Start line mismatch: ${rslintDiag.range.start.line} !== ${tsDiag.line}`,
      );
    }
    if (tsDiag.endLine) {
      assert(
        rslintDiag.range.end.line === tsDiag.endLine,
        `End line mismatch: ${rslintDiag.range.end.line} !== ${tsDiag.endLine}`,
      );
    }
    if (tsDiag.column) {
      assert(
        rslintDiag.range.start.column === tsDiag.column,
        `Start column mismatch: ${rslintDiag.range.start.column} !== ${tsDiag.column}`,
      );
    }
    if (tsDiag.endColumn) {
      assert(
        rslintDiag.range.end.column === tsDiag.endColumn,
        `End column mismatch: ${rslintDiag.range.end.column} !== ${tsDiag.endColumn}`,
      );
    }
  }
}

interface RuleTesterOptions {
  languageOptions?: {
    parserOptions?: {
      project?: string;
      tsconfigRootDir?: string;
    };
  };
}
export class RuleTester {
  options: RuleTesterOptions;
  constructor(options: RuleTesterOptions) {
    this.options = options;
  }
  public run(
    ruleName: string,
    cases: {
      valid: (
        | string
        | { code: string; options?: any; only?: boolean; skip?: boolean }
      )[];
      invalid: {
        code: string;
        errors: any[];
        options?: any;
        only?: boolean;
        skip?: boolean;
      }[];
    },
  ) {
    describe(ruleName, () => {
      let cwd =
        this.options.languageOptions?.parserOptions?.tsconfigRootDir ||
        process.cwd();
      const config = path.resolve(cwd, './rslint.json');
      let virtual_entry = path.resolve(cwd, 'src/virtual.ts');
      // test whether case has only
      let hasOnly =
        cases.valid.some(x => {
          if (typeof x === 'object' && x.only) {
            return true;
          } else {
            return false;
          }
        }) || cases.invalid.some(x => x.only);
      test('valid', async () => {
        for (const validCase of cases.valid) {
          if (typeof validCase === 'object' && validCase.skip) {
            continue;
          }
          if (hasOnly) {
            if (typeof validCase === 'string') {
              continue;
            }
            if (!validCase.only) {
              continue;
            }
          }
          const code =
            typeof validCase === 'string' ? validCase : validCase.code;

          const options =
            typeof validCase === 'string' ? [] : validCase.options || [];

          // workaround for this hardcoded path https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/no-floating-promises.test.ts#L712
          if (Array.isArray(options)) {
            for (const opt of options) {
              if (Array.isArray(opt.allowForKnownSafeCalls)) {
                for (const item of opt.allowForKnownSafeCalls) {
                  if (item.path) {
                    item.path = virtual_entry;
                  }
                }
              }
            }
          }
          const diags = await lint({
            config,
            workingDirectory: cwd,
            fileContents: {
              [virtual_entry]: code,
            },
            ruleOptions: {
              [ruleName]: options,
            },
          });

          assert(
            diags.diagnostics?.length === 0,
            `Expected no diagnostics for valid case, but got: ${JSON.stringify(diags)}`,
          );
        }
      });
      test('invalid', async t => {
        for (const item of cases.invalid) {
          const {
            code,
            errors,
            only = false,
            skip = false,
            options = [],
          } = item;
          if (skip) {
            continue;
          }
          if (hasOnly && !only) {
            continue;
          }
          const diags = await lint({
            config,
            workingDirectory: cwd,
            fileContents: {
              [virtual_entry]: code,
            },
            ruleOptions: {
              [ruleName]: options,
            },
          });
          expect(diags).toMatchSnapshot();

          assert(
            diags.diagnostics?.length > 0,
            `Expected diagnostics for invalid case`,
          );
          checkDiagnosticEqual(diags.diagnostics, errors);
        }
      });
    });
  }
}

/**
 * Simple no-op tag to mark code samples as "should not format with prettier"
 *   for the plugin-test-formatting lint rule
 */
export function noFormat(raw: TemplateStringsArray, ...keys: string[]): string {
  return String.raw({ raw }, ...keys);
}
