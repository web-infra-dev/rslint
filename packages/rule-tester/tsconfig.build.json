{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "noEmit": false
  },
  "include": ["src/**/*"],
  "exclude": ["dist", "node_modules"]
}
