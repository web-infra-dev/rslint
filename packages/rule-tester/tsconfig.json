{
  "extends": "../../tsconfig.spec.json",
  "include": ["src/**/*", "tests/**/*"],
  "exclude": ["dist", "node_modules"]
}
