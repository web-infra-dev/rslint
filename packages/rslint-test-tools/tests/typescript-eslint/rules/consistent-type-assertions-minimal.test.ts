import { RuleTester } from '@typescript-eslint/rule-tester';
import { getFixturesRootDir } from '../RuleTester.ts';

const rootPath = getFixturesRootDir();
const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootPath,
    },
  },
});

// Type declarations for test cases
declare class Generic<T> {
  constructor();
}
declare type Foo = any;
declare type A = any;

ruleTester.run('consistent-type-assertions', {
  valid: [
    {
      code: 'const x = new Generic<number>() as Foo;',
      options: [
        {
          assertionStyle: 'as',
          objectLiteralTypeAssertions: 'allow',
        },
      ],
    },
    {
      code: 'const x = b as A;',
      options: [
        {
          assertionStyle: 'as',
          objectLiteralTypeAssertions: 'allow',
        },
      ],
    },
    // Test that type declarations are not flagged by this rule
    {
      code: 'type T = { x: number };',
      options: [
        {
          assertionStyle: 'as',
          objectLiteralTypeAssertions: 'allow',
        },
      ],
    },
  ],
  invalid: [
    {
      code: 'const x = new Generic<number>() as Foo;',
      errors: [
        {
          line: 1,
          messageId: 'angle-bracket',
        },
      ],
      options: [{ assertionStyle: 'angle-bracket' }],
    },
    {
      code: 'const x = b as A;',
      errors: [
        {
          line: 1,
          messageId: 'angle-bracket',
        },
      ],
      options: [{ assertionStyle: 'angle-bracket' }],
    },
  ],
});
