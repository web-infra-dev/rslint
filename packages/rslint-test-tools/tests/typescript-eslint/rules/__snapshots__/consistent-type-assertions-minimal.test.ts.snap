// Rstest Snapshot v1

exports[`consistent-type-assertions > invalid 1`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Use '<Foo>' instead of 'as Foo'.",
      "messageId": "as-assertion",
      "range": {
        "end": {
          "column": 39,
          "line": 1,
        },
        "start": {
          "column": 11,
          "line": 1,
        },
      },
      "ruleName": "consistent-type-assertions",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`consistent-type-assertions > invalid 2`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Use '<A>' instead of 'as A'.",
      "messageId": "as-assertion",
      "range": {
        "end": {
          "column": 17,
          "line": 1,
        },
        "start": {
          "column": 11,
          "line": 1,
        },
      },
      "ruleName": "consistent-type-assertions",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;
